teps/0140-producing-results-in-matrix.md

```diff
@@ -0,0 +1,702 @@
+	PipelineTask   *v1.PipelineTask
+	ResolvedTask   *resources.ResolvedTask
+	ResultsCache   map[string][]string
+}
+```
+## Design
+
+### Support String Results produced by a matrix, Aggregated as An Array during Fanning Out - Referencing using indexes
+
+Since the ordering of a fanned out matrix is predeterministic due to sorting the combinations, we can reference results produced from different TaskRuns using indexing using the syntax  `$(tasks.<pipelinetask-name>.results.<result-name>[INDEX])`.  This approach will work for any matrix regardless if it has matrix params and/or include params.  A con to this approach is that users will need to know in advance how the order of the TaskRuns is determined and figure out what the order will be ahead of time if users need to reference or filter for a specific result. At the pipeline definition time, the parameters are unknown. Another con is that it might become exponentially more challenging as you scale with more combinations, making it less user friendly.
+In the future, we could look into supporting adding the combinations as part of the result so the task consuming the result can map which combinations produced which result.
+
+### 1. Building Images
+In the example below, a user is able to produce and deploy images using specific combinations of images and dockerfiles using the task  ”matrix-emitting-results” which produces the result ”IMAGE-DIGEST”, which is used by another matrixed PipelineTask “task-deploy-images” to deploy the images.
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  generateName: matrix-emitting-results
+spec:
+  serviceAccountName: "default"
+  pipelineSpec:
+    tasks:
+      - name: matrix-emitting-results
+        matrix:
+          include:
+            - name: build-1
+              params:
+                - name: IMAGE
+                  value: image-1
+                - name: DOCKERFILE
+                  value: path/to/Dockerfile1
+            - name: build-2
+              params:
+                - name: IMAGE
+                  value: image-2
+                - name: DOCKERFILE
+                  value: path/to/Dockerfile2
+            - name: build-3
+              params:
+                - name: IMAGE
+                  value: image-3
+                - name: DOCKERFILE
+                  value: path/to/Dockerfile3
+        taskSpec:
+          params:
+            - name: IMAGE
+            - name: DIGEST
+              default: ""
+          results:
+            - name: IMAGE-DIGEST
+          steps:
+            - name: produce-image-digest
+              image: bash:latest
+              script: |
+                echo "Building image for $(params.IMAGE)"
+                echo -n "$(params.IMAGE)" | sha256sum | tee $(results.IMAGE-DIGEST.path)
+      - name: task-deploy-images
+        params:
+          - name: DIGEST
+            value: $(tasks.matrix-emitting-results.results.IMAGE-DIGEST[*])
```

Comment:
how about creating a new type for result used in matrix? If we're concerned about the type is string but changed to array later. How about just name a new type (e.g."matrix") to meet our needs

Reply:
from chatting with @EmmaMunley yesterday, it seems like the DIGEST param should actually be of type `array` so there is not type changing going on?